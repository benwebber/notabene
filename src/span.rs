//! Work with spans within the source document.
use std::ops::{Deref, Range};

use serde::{Deserialize, Serialize};

mod iter;
mod locator;
mod ranged;
mod unist;

pub(crate) use iter::SpanIterator;
pub use locator::Locator;
pub(crate) use ranged::Ranged;
pub use unist::{Point, Position};

/// A span within the source document.
///
/// Use [`Locator`] to convert a `Span` to a [`Position`].
#[derive(
    Copy, Clone, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize,
)]
pub struct Span {
    /// The start offset of the span.
    pub start: usize,
    /// The end offset of the span.
    pub end: usize,
}

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize)]
pub(crate) struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

impl Ranged<usize> for Span {
    fn range(&self) -> Range<usize> {
        (*self).into()
    }
}

impl<'a> Deref for Spanned<&'a str> {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        self.value
    }
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    /// Offset a span by `offset`.
    pub fn offset(&self, offset: usize) -> Self {
        Self {
            start: offset + self.start,
            end: offset + self.end,
        }
    }
}

impl From<Range<usize>> for Span {
    fn from(range: Range<usize>) -> Self {
        Self {
            start: range.start,
            end: range.end,
        }
    }
}

impl From<Span> for Range<usize> {
    fn from(span: Span) -> Self {
        span.start..span.end
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_offset() {
        let span = Span::new(0, 10);
        assert_eq!(span.offset(5), Span::new(5, 15));
    }

    #[test]
    fn test_range() {
        let span = Span::new(0, 10);
        assert_eq!(span.range(), span.into())
    }
}
