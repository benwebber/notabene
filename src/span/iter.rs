use super::Span;

pub struct SpanIterator<'a> {
    text: &'a str,
    cursor: usize,
}

impl<'a> SpanIterator<'a> {
    pub fn new(text: &'a str) -> Self {
        Self { text, cursor: 0 }
    }
}

impl<'a> Iterator for SpanIterator<'a> {
    type Item = Span;

    fn next(&mut self) -> Option<Self::Item> {
        let start = self.text[self.cursor..]
            .char_indices()
            .find(|(_, c)| !c.is_whitespace())
            .map(|(i, _)| self.cursor + i);
        let Some(start) = start else {
            // End of string.
            self.cursor = self.text.len();
            return None;
        };
        let end = self.text[start..]
            .char_indices()
            .find(|(_, c)| c.is_whitespace())
            .map(|(i, _)| start + i)
            .unwrap_or(self.text.len());
        self.cursor = end;
        Some(Span::new(start, end))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_span_iterator() {
        let s = "  foo  foobar  quux  ";
        let spans: Vec<Span> = SpanIterator::new(s).collect();
        assert_eq!(
            spans,
            vec![Span::new(2, 5), Span::new(7, 13), Span::new(15, 19)]
        );
    }
}
