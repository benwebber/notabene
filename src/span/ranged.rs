pub trait Ranged<T> {
    fn range(&self) -> std::ops::Range<T>;
}

impl Ranged<usize> for std::ops::Range<usize> {
    fn range(&self) -> Self {
        self.clone()
    }
}
