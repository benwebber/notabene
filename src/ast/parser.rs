//! Parse a Markdown document as an iterator of block elements.
use pulldown_cmark as md;

use super::*;
use crate::span::Span;

/// A [pulldown-cmark](https://crates.io/crates/pulldown-cmark) parser that includes text offset
/// information (for spans) and merges consecutive text nodes into one.
type MdParser<'a> = md::utils::TextMergeWithOffset<'a, md::OffsetIter<'a>>;

pub struct Parser<'a> {
    inner: MdParser<'a>,
}

impl<'a> Parser<'a> {
    pub fn new(s: &'a str) -> Self {
        let parser = md::Parser::new(s).into_offset_iter();
        let inner = md::utils::TextMergeWithOffset::new(parser);
        Self { inner }
    }
}

impl Literal {
    fn new<S: Into<Span>>(span: S) -> Self {
        Self { span: span.into() }
    }
}

impl Iterator for Parser<'_> {
    type Item = Block;

    fn next(&mut self) -> Option<Self::Item> {
        while let Some((event, range)) = self.inner.next() {
            match event {
                md::Event::Start(md::Tag::Heading { level, .. }) => {
                    let inlines = parse_inlines(&mut self.inner, md::TagEnd::Heading(level));
                    let level = match level {
                        md::HeadingLevel::H1 => 1,
                        md::HeadingLevel::H2 => 2,
                        md::HeadingLevel::H3 => 3,
                        md::HeadingLevel::H4 => 4,
                        md::HeadingLevel::H5 => 5,
                        md::HeadingLevel::H6 => 6,
                    };
                    // Remove trailing newline.
                    let span = Span::new(range.start, range.end - 1);
                    let heading = Heading {
                        span,
                        level,
                        inlines,
                    };
                    return Some(Block::Heading(heading));
                }
                md::Event::Start(md::Tag::Paragraph) => {
                    let lit = Literal::new(range);
                    return Some(Block::Paragraph(lit));
                }
                md::Event::Start(md::Tag::List(None)) => {
                    let items = parse_items(&mut self.inner);
                    let list = List {
                        span: range.into(),
                        items,
                    };
                    return Some(Block::List(list));
                }
                _ => {}
            }
        }
        None
    }
}

fn parse_inlines(parser: &mut MdParser, until: md::TagEnd) -> Vec<Inline> {
    let mut inlines = vec![];
    while let Some((event, span)) = parser.next() {
        match event {
            md::Event::Start(md::Tag::Link { dest_url, .. }) => {
                let content_span = read_span_until(parser, md::TagEnd::Link);
                let content = Literal::new(content_span);
                let target: String = dest_url.into();
                let link = Link {
                    span: span.into(),
                    content,
                    target,
                };
                inlines.push(Inline::Link(link))
            }
            md::Event::Text(_) => {
                let span = Span::from(span);
                inlines.push(Inline::Literal(Literal::new(span)));
            }
            md::Event::End(tag) if tag == until => break,
            _ => {}
        }
    }
    inlines
}

fn parse_items(parser: &mut MdParser) -> Vec<Literal> {
    let mut items = vec![];
    while let Some((event, _)) = parser.next() {
        match event {
            md::Event::Start(md::Tag::Item) => {
                let item_span = read_span_until(parser, md::TagEnd::Item);
                items.push(Literal::new(item_span));
            }
            md::Event::End(md::TagEnd::List(false)) => break,
            _ => {}
        }
    }
    items
}

fn read_span_until(parser: &mut MdParser, until: md::TagEnd) -> Span {
    let mut span = Span::default();
    for (event, span_) in parser.by_ref() {
        match event {
            md::Event::End(tag) if tag == until => {
                break;
            }
            _ => {
                if span.start == 0 {
                    span.start = span_.start
                }
                span.end = span_.end
            }
        }
    }
    span
}

#[cfg(test)]
mod tests {

    use super::*;

    macro_rules! snapshot {
        ($s:literal) => {
            let blocks: Vec<Block> = Parser::new(&$s.trim()).collect();
            insta::assert_yaml_snapshot!(blocks);
        };
    }

    #[test]
    fn test_parse_h1() {
        snapshot!("# Changelog");
        snapshot!(
            "
Changelog
=========
        "
        );
    }

    #[test]
    fn test_parse_h2() {
        snapshot!(
            "
## [Unreleased]

[Unreleased]: https://example.org/
        "
        );
        snapshot!(
            "
[Unreleased]
------------

[Unreleased]: https://example.org/
        "
        );
        snapshot!("## [Unreleased](https://example.org/)");
        snapshot!(
            "
[Unreleased](https://example.org/)
----------------------------------
"
        );
        snapshot!(
            "
## [1.6.8] - 2038-01-19 [YANKED]

[1.6.8]: https://example.org/
"
        );
        snapshot!(
            "
[1.6.8] - 2038-01-19 [YANKED]
-----------------------------

[1.6.8]: https://example.org/
"
        );
        snapshot!("## [1.6.8](https://example.org/) - 2038-01-19 [YANKED]");
        snapshot!(
            "
[1.6.8](https://example.org/) - 2038-01-19 [YANKED]
---------------------------------------------------
"
        );
    }

    #[test]
    fn test_parse_h3() {
        snapshot!("### Added");
    }

    #[test]
    fn test_parse_paragraph() {
        snapshot!(
            "
Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.

Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.
        ");
    }

    #[test]
    fn test_parse_list() {
        snapshot!(
            "
* Lorem ipsum dolor sit amet
* Consectetur adipiscing elit
* Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
        "
        );
    }
}
