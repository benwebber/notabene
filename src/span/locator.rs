use std::cmp::Ordering;

use super::{Point, Position, Ranged, Span};

/// Locates spans in the source document.
pub struct Locator<'a> {
    source: &'a str,
    line_spans: Vec<Span>,
}

impl<'a> Locator<'a> {
    pub fn new(source: &'a str) -> Self {
        let mut line_spans = Vec::new();
        let mut start = 0;
        for (end, _) in source.match_indices('\n') {
            line_spans.push(Span::new(start, end));
            start = end + 1;
        }
        if start <= source.len() {
            // No trailing newline.
            line_spans.push(Span::new(start, source.len()))
        }
        Self { source, line_spans }
    }

    /// Return the unist Point for the given offset.
    fn point(&self, offset: usize) -> Point {
        if self.line_spans.is_empty() {
            return Point::default();
        }
        let cmp = |span: &Span| {
            if offset < span.start {
                Ordering::Greater
            } else if offset > span.end {
                Ordering::Less
            } else {
                Ordering::Equal
            }
        };
        let i = match self.line_spans.binary_search_by(cmp) {
            Ok(i) => i,
            Err(i) => i
                .saturating_sub(1)
                .min(self.line_spans.len().saturating_sub(1)),
        };
        Point::new(
            i + 1,
            offset.saturating_sub(self.line_spans[i].start) + 1,
            offset,
        )
    }

    /// Return the unist Position for the given span.
    pub fn position<R: Ranged<usize>>(&self, ranged: &R) -> Position {
        let range = ranged.range();
        Position::new(self.point(range.start), self.point(range.end))
    }

    /// Return the contents of line number `line`.
    pub(crate) fn line(&self, line: usize) -> &'a str {
        &self.source[self.line_spans[line.saturating_sub(1)].range()]
    }

    /// Return the number of lines in the source.
    pub(crate) fn lines(&self) -> usize {
        self.line_spans.len()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::span::SpanIterator;

    #[test]
    fn test_position() {
        let s = "foo

bar

quux";
        let locator = Locator::new(s);
        let spans: Vec<Span> = SpanIterator::new(s).collect();
        assert_eq!(
            locator.position(&spans[0]),
            Position::new(Point::new(1, 1, 0), Point::new(1, 4, 3)),
        );
        assert_eq!(
            locator.position(&spans[1]),
            Position::new(Point::new(3, 1, 5), Point::new(3, 4, 8)),
        );
        assert_eq!(
            locator.position(&spans[2]),
            Position::new(Point::new(5, 1, 10), Point::new(5, 5, 14)),
        );
    }

    #[test]
    fn test_line() {
        let s = "foo bar baz\nfoobar foobaz";
        let locator = Locator::new(s);
        assert_eq!(locator.line(2), "foobar foobaz");
    }
}
