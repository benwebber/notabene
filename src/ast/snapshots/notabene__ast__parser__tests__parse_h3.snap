---
source: src/ast/parser.rs
expression: blocks
---
- Heading:
    span:
      start: 0
      end: 8
    level: 3
    inlines:
      - Literal:
          span:
            start: 4
            end: 9
